tmva/tmva/inc/TMVA/RTensor.hxx

```diff
@@ -0,0 +1,223 @@
+#ifndef ROOT_TMVA_RTENSOR
+#define ROOT_TMVA_RTENSOR
+
+#include "ROOT/RVec.hxx"
+
+#include <sstream>
+#include <cstdint>
+
+namespace TMVA {
+namespace Experimental {
+
+/// Memory order types
+struct MemoryOrder {
+   static const uint8_t RowMajor = 0;
+   static const uint8_t ColumnMajor = 1;
+};
+
+namespace Internal {
+
+/// Get size of tensor from shape vector
+inline size_t GetSizeFromShape(const std::vector<size_t> &shape)
+{
+   if (shape.size() == 0)
+      return 0;
+   size_t size = 1;
+   for (auto &s : shape)
+      size *= s;
+   return size;
+}
+
+/// Compute strides from shape vector.
+/// This information is needed for the multi-dimensional indexing. See here:
+/// https://en.wikipedia.org/wiki/Row-_and_column-major_order
+/// https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.strides.html
+inline std::vector<size_t> ComputeStrides(const std::vector<size_t> &shape, uint8_t memoryOrder)
+{
+   const auto size = shape.size();
+   std::vector<size_t> strides(size);
+   if (memoryOrder == MemoryOrder::RowMajor) {
+      for (size_t i = 0; i < size; i++) {
+         if (i == 0) {
+            strides[size - 1 - i] = 1;
+         } else {
+            strides[size - 1 - i] = strides[size - 1 - i + 1] * shape[size - 1 - i + 1];
+         }
+      }
+   } else if (memoryOrder == MemoryOrder::ColumnMajor) {
+      for (size_t i = 0; i < size; i++) {
+         if (i == 0) {
+            strides[i] = 1;
+         } else {
+            strides[i] = strides[i - 1] * shape[i - 1];
+         }
+      }
+   } else {
+      std::stringstream ss;
+      ss << "Memory order type " << memoryOrder << " is not known for calculating the strides.";
+      throw std::runtime_error(ss.str());
+   }
+   return strides;
+}
+
+} // namespace TMVA::Experimental::Internal
+
+/// RTensor class
+template <typename T>
+class RTensor {
+public:
+   // Constructors
+   RTensor(const std::vector<size_t> &shape, uint8_t memoryOrder = MemoryOrder::RowMajor);
+   RTensor(T *data, const std::vector<size_t> &shape, uint8_t memoryOrder = MemoryOrder::RowMajor);
+
+   // Access elements
+   T &At(const std::vector<size_t> &idx);
+   template <typename... Idx>
+   T &operator()(Idx... idx);
+
+   // Shape modifications
+   void Reshape(const std::vector<size_t> &shape);
+
+   // Get properties of container
+   size_t GetSize() const { return fSize; }                    // Return size of contiguous memory
+   std::vector<size_t> GetShape() const { return fShape; }     // Return shape
+   std::vector<size_t> GetStrides() const { return fStrides; } // Return strides
+   T *GetData() { return fData.data(); }                       // Return pointer to data
+   ROOT::VecOps::RVec<T> &GetDataAsVec() { return fData; }     // Return data as reference to vector
+   uint8_t GetMemoryOrder() { return fMemoryOrder; }           // Return memory order
+
+private:
+   size_t fSize;                 // Size of contiguous memory
```

Comment:
why do you need to store both `fSize` and `fShape`? (also `fSize` is also stored in `fData`)